core/src/replay_stage.rs

```diff
@@ -1469,11 +1461,104 @@ impl ReplayStage {
             vote_signatures.clear();
         }
 
-        let _ = cluster_info.send_vote(
-            &vote_tx,
-            crate::banking_stage::next_leader_tpu(cluster_info, poh_recorder),
+        Some(vote_tx)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn refresh_last_vote(
+        tower: &mut Tower,
+        cluster_info: &ClusterInfo,
+        heaviest_bank_on_same_fork: &Bank,
+        poh_recorder: &Mutex<PohRecorder>,
+        my_latest_landed_vote: Slot,
+        vote_account_pubkey: &Pubkey,
+        authorized_voter_keypairs: &[Arc<Keypair>],
+        vote_signatures: &mut Vec<Signature>,
+        has_new_vote_been_rooted: bool,
+        last_vote_refresh_time: &mut Instant,
+    ) {
+        let last_voted_slot = tower.last_voted_slot();
+        if last_voted_slot.is_none() {
+            return;
+        }
+
+        // Refresh the vote if our latest vote hasn't landed, and the recent blockhash of the
+        // last attempt at a vote transaction has expired
+        let last_voted_slot = last_voted_slot.unwrap();
+        if my_latest_landed_vote >= last_voted_slot
+            || heaviest_bank_on_same_fork
+                .check_hash_age(&tower.last_vote_tx_blockhash(), MAX_PROCESSING_AGE)
+                .unwrap_or(false)
+            // In order to avoid voting on multiple forks all past MAX_PROCESSING_AGE that don't
+            // include the last voted blockhash
+            || last_vote_refresh_time.elapsed().as_millis() < 5000
+        {
+            return;
+        }
+
+        // TODO: check the timestamp in this vote is correct, i.e. it shouldn't
+        // have changed from the original timestamp of the vote.
```

Comment:
@CriesofCarrots just wanted to check that this is safe for the time adjustments. Revoting for the same slot with a different blockhash, but the `Vote.timestamp` remains unchanged

Reply:
Yes. As long as both the slot and the timestamp are the same as in the previous vote, the Vote program timestamp-check code should be happy.